support extraNames in x509Subject. Fixes #4424

yongk802 opened the pull request:
### Pull Request Motivation

Some of the X509Subject fields are not supported like GN and DC and it would be nice to have these officially supported instead of using workarounds.

### Kind

feature

### Release Note

```release-note
Support extra Subject Name fields as part of the CertificateSpec in a generic manner by allowing the caller to pass in an asn1 OID string and value. 
```

yongk802 commented:
/assign @wallrj

yongk802 commented:
/test pull-cert-manager-make-test

yongk802 commented:
/retest

yongk802 commented:
/retest

yongk802 commented:
@wallrj Hi Richard, is there any interest in taking a look at this PR? Thanks!

yongk802 commented on pkg/controller/certificates/util.go:
```diff
@@ -178,6 +182,29 @@ func RequestMatchesSpec(req *cmapi.CertificateRequest, spec cmapi.CertificateSpe
+		return true
+	}
+
+	for _, requestAttributeTypeAndValue := range requestNames {
```
Just wanted to add a comment here to have someone take a second look at this part as it's probably the most critical piece of logic.

munnerz replied:
Please write some unit tests for this function - I feel like there's a number of cases where this may not return as expected (it seems like you're returning as soon as the first matching extraName is found, but that doesn't account for any of the _other_ extraNames if there are multiple).

Also, should the _order_ of these matter?

yongk802 replied:
That is true - it's probably more correct to do:

```
if subjectExtraName.Value != requestAttributeTypeAndValue.Value {
return false
}
```

I tried adding some unit tests in a new file, but the bazel build failed, so I deleted them. I did add them to some sort of bazel manifest where it seems to list all of the files in that folder/module because it looked like that's what one was supposed to do, but it still failed. I'll add back the unit tests along with more of them and ask for help in the slack channel if it keeps failing (unless you know off-hand why that might be).

munnerz replied:
Did you run `./hack/update-bazel.sh` after adding the new file? 😄

yongk802 replied:
I did, I actually ran both update-bazel.sh and update-all.sh. I tried update-all after update-bazel didn't do the trick. I'll try again when I make these updates. There was some weird build stuff going on at the time.

SgtCoDFish commented:
Sorry, just want to keep you in the loop: I'm really busy currently and I'm on holiday next week, so I can't take a look any time soon. It might be worth asking again in #cert-manager-dev on slack for a review.

Wish I had more time but I just don't! Sorry!

munnerz commented on deploy/crds/crd-certificates.yaml:
```diff
@@ -229,6 +229,18 @@ spec:
+                            description: The type. This should be an asn1.ObjectIdentifier string, e.g., "2.5.4.42"
+                            type: string
+                          value:
+                            description: The value. This should be an IA5String
```
What's an IA5String? Can you provide some examples, references etc in this documentation?

yongk802 replied:
Basically, it's ASCII (which is the US implementation of IA5), but with a couple of small differences. Here's the best article I could find to explain it. https://www.zytrax.com/tech/ia5.html. I can put a comment in here that says something to that effect.

munnerz commented on deploy/crds/crd-certificates.yaml:
```diff
@@ -229,6 +229,18 @@ spec:
                       items:
                         type: string
+                    extraNames:
+                      description: ExtraNames is a list of pkix.ExtraNames to be set on the Certificate.
```
From a user's perspective, what is a pkix.ExtraNames? Is there a more general (ie not go specific) term to use here?

yongk802 replied:
We could name it something like SubjectExtraFields? I named it this way so it was clear how it maps to the underlying golang pkix library. Otherwise, someone trying to understand the code would need to know it changes from SubjectExtraNames to pkix.ExtraNames. I'm not opposed to it, but I'm not sure if that makes it clearer or not.

munnerz commented on internal/apis/certmanager/v1alpha2/types_certificate.go:
```diff
@@ -275,6 +275,15 @@ type X509Subject struct {
 	// +optional
 	SerialNumber string `json:"serialNumber,omitempty"`
+	// Extra names to be used on the Certificate.
+	ExtraNames []ExtraName `json:"extraNames,omitempty"`
```
Should this be under the 'subject' field? Also curious as to how this will relate to the other 'RawSubject' proposal that's being looked into too 👀

yongk802 replied:
I'm not aware of the "RawSubject" proposal. Is there a link I can review about that? I just want to know why that's needed as I don't see a need for it if we're allowing users to use OIDs directly. Also, that sounds like the validation would be even trickier than using this method.

munnerz replied:
https://github.com/cert-manager/cert-manager/pull/5002 there you go 😄 I think a key request there is around how the subject is actually rendered, and the ordering of the components of it

yongk802 replied:
Ok, that makes sense. I think there will be some merge conflicts, but if the rawsubject is provided, then it should skip this ExtraNames field. It's interesting that the ordering matters and not just the ordering, but also how it's grouped. It sounds like that would still be useful for those wanting complete control. I also think there should be some description for what that field is in the PR and how it works. I'll leave a comment about it.

yongk802 replied:
On second thought... if it is empty, i think it should show as empty, so yes, deleting both omitEmpty annotations is probably correct.

spockz replied:
>Should this be under the 'subject' field?

I think so as this part of the subject, just like `ExtraNames` in are under the `Name` in the pkix package.

yongk802 replied:
Agreed. At first, I tried putting it where CommonName was, at the top, but while coding it this way, it didn't feel right. 

To give context, I tried this both ways, both under Subject and under CertificateSpec and now I feel If anything, CommonName being a top-level field is weird. I get it that it's kind of a first-class citizen in the world of certificates, but it really should be part of the subject as it's not truly a top-level field when you look at a certificate.

I had to have some code to map between the "Subject" of cert-manager into the "Subject" of pkix.Name and back because the pkix library code's structure follows the layout of a human-readable certificate where all "Subject" fields are inside pkix.Name.

munnerz commented on internal/apis/certmanager/v1alpha2/types_certificate.go:
```diff
@@ -275,6 +275,15 @@ type X509Subject struct {
+	// The type. This should be an asn1.ObjectIdentifier string, e.g., "2.5.4.42"
+	Type string `json:"type,omitempty"`
+	// The value. This should be an IA5String
+	Value string `json:"value,omitempty"`
```
The 'type' and 'value' fields are presumably not optional, so we shouldn't have 'omitempty' here 😊

Also there should be validation on both of these fields - eg the type can be checked to ensure it's a valid identifier, and then I'm uncertain precisely what validation we can perform on the value, but at the very least a maxLength check should be in place.

yongk802 replied:
Hm, I do agree that the Type shouldn't be empty. However, I don't think anything prevents one from having a type with an empty value. I'm not sure what the use-case would be, but technically, null values are allowed for values because null is a valid IA5 character.

yongk802 replied:
On second thought... if it is empty, i think it should show as empty, so yes, deleting both omitEmpty annotations is probably correct.

munnerz commented on pkg/util/pki/keyusage.go:
```diff
@@ -148,3 +150,23 @@ func buildASN1KeyUsageRequest(usage x509.KeyUsage) (pkix.Extension, error) {
+
+// Allow users of the API to pass in an OID string and we will convert to an asn1.ObjectIdentifier
+// https://datatracker.ietf.org/doc/html/rfc4519.html
+func ToAsn1Oid(oidString string) (asn1.ObjectIdentifier, error) {
```
We should perform all this kind of validation with a regex pattern on the CRD to avoid having any unparseable/invalid data 😊

yongk802 replied:
Hm - that's an interesting idea. I didn't even know you could do that. I just looked it up and it looks like you can use the "pattern" field to do this? I'll give it a shot, thanks.

munnerz commented on test/unit/gen/certificate.go:
```diff
@@ -80,6 +80,26 @@ func SetCertificateEmails(emails ...string) CertificateModifier {
 	}
 }
 
+func SetCertificateGivenName(givenName string) CertificateModifier {
```
Why 'GivenName' here instead of 'ExtraName'?

yongk802 replied:
I thought it would be handy to show how to set a specific ExtraName field, but yeah, it does feel weird. Perhaps it should be more generic and be 2 methods. One that sets an ExtraName field, and in the test, create an ExtraName field and pass it in there.

munnerz reviewed (commented):
I've done an initial pass/review - it'd also be great to see some extra tests to ensure that CSRs generated by the requestmanager controller actually contain these fields in the x509 CSR that it creates 😊

I'm also still a bit nervous about exposing such raw values here - could someone use this to set a dnsName by using the OID of dnsSANs here, etc?

yongk802 commented:
@munnerz I pushed up most of the changes discussed. Thanks a lot for the suggestions and review!

I added some tests and fixed some of that code. I left the name ExtraNames field as I couldn't really think of a better name than what the go library is using, but I did update the comment in an attempt to try to make it more clear. 

I gave some thought to what happens if you pass the same OID multiple times with different values and the ordering. I suppose we could try to be clever and maybe copy the array and remove items from it as we find them... or some sort of pattern like that, but I thought that might be too fancy, or it might be I'm just too lazy to do something like that. I did put a unit test in there with a TODO that says it would be nice to have. Passing an array with the same OID and different values is a foreign concept to me - though I do see it seems to have a use-case according to the RawSubject PR.

One more thing I need to check on is what happens if you provide an OID for say CommonName in the ExtraName object. Does it over-ride the CN set?

> I've done an initial pass/review - it'd also be great to see some extra tests to ensure that CSRs generated by the requestmanager controller actually contain these fields in the x509 CSR that it creates 😊
> 
> I'm also still a bit nervous about exposing such raw values here - could someone use this to set a dnsName by using the OID of dnsSANs here, etc?

This is an interesting point so I did some testing using CommonName. Yes, it seems one could use that OID (in this case, "2.5.4.3") to overwrite the CommonName passed in through the original CertificateSpec. This has the effect of the CertificateSpec having the original CommonName and CSR having the one from ExtraNames. This would cause validation to fail. 

I don't know if that's something we should guard against by trying to strip out all officially supported fields from ExtraNames passed in or just warn the user that is the behavior. I also don't see in the code where the CSR is generated so I can do that sort of filtering. 

Thoughts?

yongk802 commented:
The bazel thing failed again. I definitely ran hacks/update-bazel.sh. I am almost positive that if I remove that file I added, "keyusage_test.go" this error goes away as that's what fixed it for me before, as perplexing as that is.

yongk802 commented:
> I've done an initial pass/review - it'd also be great to see some extra tests to ensure that CSRs generated by the requestmanager controller actually contain these fields in the x509 CSR that it creates 😊
> 
> I'm also still a bit nervous about exposing such raw values here - could someone use this to set a dnsName by using the OID of dnsSANs here, etc?


It's actually interesting what the golang lib does here:

```
// ToRDNSequence converts n into a single RDNSequence. The following
// attributes are encoded as multi-value RDNs:
//
//  - Country
//  - Organization
//  - OrganizationalUnit
//  - Locality
//  - Province
//  - StreetAddress
//  - PostalCode
//
// Each ExtraNames entry is encoded as an individual RDN.
func (n Name) ToRDNSequence() (ret RDNSequence) {
	ret = n.appendRDNs(ret, n.Country, oidCountry)
	ret = n.appendRDNs(ret, n.Province, oidProvince)
	ret = n.appendRDNs(ret, n.Locality, oidLocality)
	ret = n.appendRDNs(ret, n.StreetAddress, oidStreetAddress)
	ret = n.appendRDNs(ret, n.PostalCode, oidPostalCode)
	ret = n.appendRDNs(ret, n.Organization, oidOrganization)
	ret = n.appendRDNs(ret, n.OrganizationalUnit, oidOrganizationalUnit)
	if len(n.CommonName) > 0 {
		ret = n.appendRDNs(ret, []string{n.CommonName}, oidCommonName)
	}
	if len(n.SerialNumber) > 0 {
		ret = n.appendRDNs(ret, []string{n.SerialNumber}, oidSerialNumber)
	}
	for _, atv := range n.ExtraNames {
		ret = append(ret, []AttributeTypeAndValue{atv})
	}

	return ret
}

func (n Name) appendRDNs(in RDNSequence, values []string, oid asn1.ObjectIdentifier) RDNSequence {
	if len(values) == 0 || oidInAttributeTypeAndValue(oid, n.ExtraNames) {
		return in
	}

	s := make([]AttributeTypeAndValue, len(values))
	for i, value := range values {
		s[i].Type = oid
		s[i].Value = value
	}

	return append(in, s)
}

func oidInAttributeTypeAndValue(oid asn1.ObjectIdentifier, atv []AttributeTypeAndValue) bool {
	for _, a := range atv {
		if a.Type.Equal(oid) {
			return true
		}
	}
	return false
}

```

If I'm not mistaken, it looks like it will append values unless the value already exists, then it just returns itself - at least for the known fields is has like CommonName. So it kind of keeps itself from overwriting values in the RDNSequence. However, when it gets to the ExtraNames, it doesn't care and just throws everything in, so it would overwrite anything in their if you've provided a clashing OID. At least that's how I'm interpreting this code. I don't know if there's any precedent for this behavior, but I found this doc for an Erlang project and it basically describes exactly what I'd like to do here. 

That being said, maybe it's enough to just document the behaviors and give fair warning?

https://hexdocs.pm/x509/X509.RDNSequence.html

```
Note that this module implements a commonly used subset of RDNSequence values. It supports only a single attribute type/value pair for each element in the sequence, and it implements the attribute types specified in RFC5280, section 4.1.2.4, with a few extensions from LDAP:

countryName (C)
organizationName (O)
organizationalUnitName (OU)
dnQualifier
stateOrProvinceName (ST)
commonName (CN)
serialNumber
localityName (L)
title
name
surname (SN)
givenName (GN)
initials
pseudonym
generationQualifier
domainComponent (DC)
emailAddress (E)
This module encodes values as UTF8String where possible, or else PrintableString/IA5String where required.
```

yongk802 commented on pkg/controller/certificates/issuing/issuing_controller_test.go:
```diff
@@ -71,6 +76,8 @@ func TestIssuingController(t *testing.T) {
 		gen.SetCertificateRevision(1),
 		gen.SetCertificateNextPrivateKeySecretName(nextPrivateKeySecretName),
+		gen.SetCertificateCommonName("commonname"),
+		gen.SetCertificateExtraName(givenName),
```
So... I thought with my change here I was testing that my change was working and adding these fields, but writing more tests elsewhere in the code (ca_test.go), now I'm not so sure. 

Taking a hard look, I don't actually understand what maps the OIDs into the certificate so that something knows when it sees "2.4.5.43" or whatever it should display it as "GN" without some sort of actual mapping. I had assumed it was some algorithm based off the number array, but now I no longer believe that to be the case.

I was looking. at the other PR you had mentioned and it's trying to achieve something very similar and I see it is doing a lookup in the mapping, similar to what the pkix go lib does, so now I'm starting to think I actually am missing a big piece (that mapping). 

I am going to do some playing around with the other PR as I now think it might be best to build on that and try to extend it and we will need to add explicit fields (givenname, domaincomponent, etc), which I was trying really hard to avoid.

yongk802 commented:
@munnerz I think I have this PR in good condition now. I added some tests and all tests are passing. 

The basic idea is that you pass in the ExtraNames, you then get those added to the CSR and if signed, you get a certificate back where you can see everything back in the Subject.Names field. For my use case, I'm passing the CSR to another entity using an external provider (EJBCA) I created based off the aws/google examples.

It might be a little confusing, but due to how the GO pkix lib handles Names vs ExtraNames during marshaling and unmarshaling from DER, I think this is the best we can do. I basically throw all the ExtraNames into Names right before signing as a workaround for the below to get back those fields in Names in the parsed Certificate object. This means the developer will need to know to look in Names vs ExtraNames for those fields, which is a little confusing, but seems like the best option without re-writing parts of the pkix lib. Here is the documentation for those fields, which is why I decided to take this approach:

```
	// Names contains all parsed attributes. When parsing distinguished names,
	// this can be used to extract non-standard attributes that are not parsed
	// by this package. When marshaling to RDNSequences, the Names field is
	// ignored, see ExtraNames.
	Names []AttributeTypeAndValue

	// ExtraNames contains attributes to be copied, raw, into any marshaled
	// distinguished names. Values override any attributes with the same OID.
	// The ExtraNames field is not populated when parsing, see Names.
	ExtraNames []AttributeTypeAndValue
```

I also manually checked the csr and certs in pem format using openssl and verified the additional fields show up as expected there. Openssl must have facilities to read the OIDs and map them to the short names (i.e., 2.5.4.3 -> GN.

Of course, this all assumes the code is doing what I think it's doing., which I've spent a lot of time studying, so I'm hoping I am right.

Let me know your thoughts. Thanks!

yongk802 commented:
/retest

spockz commented on pkg/controller/certificates/util.go:
```diff
@@ -167,6 +168,9 @@ func RequestMatchesSpec(req *cmapi.CertificateRequest, spec cmapi.CertificateSpe
 	if !util.EqualKeyUsagesUnsorted(req.Spec.Usages, spec.Usages) {
 		violations = append(violations, "spec.usages")
 	}
+	if !ExtraNameFieldsEqual(x509req.Subject.Names, spec.Subject.ExtraNames) {
```
Could this be a DeepEquals?

yongk802 replied:
Unfortunately, no. There's some slightly more complex logic here for equals.

jakexks commented:
/milestone v1.9

jakexks commented:
I think the functionality you require is exposed by https://github.com/cert-manager/cert-manager/pull/5002 so I will close this issue. Let me know if it's insufficient
